src/cli.rs

```diff
@@ -102,96 +92,68 @@ pub fn init() {
+    let log_path = MANAGER
+        .read()
+        .unwrap()
+        .clap_matches
+        .log_path
+        .clone()
+        .expect("Clap arg \"log-path\" should always be \"Some(_)\" because of the default value.");
 
     shellexpand::full(&log_path)
         .expect("Failed to expand path")
         .to_string()
 }
 
-#[instrument(level = "debug")]
-pub fn tcp_client_endpoints() -> Vec<String> {
-    get_endpoint_with_kind("tcpc")
-}
-
-#[instrument(level = "debug")]
-pub fn file_server_endpoints() -> Vec<String> {
-    get_endpoint_with_kind("file")
-}
-
-#[instrument(level = "debug")]
-pub fn tcp_server_endpoints() -> Vec<String> {
-    get_endpoint_with_kind("tcps")
-}
-
-#[instrument(level = "debug")]
-pub fn udp_client_endpoints() -> Vec<String> {
-    get_endpoint_with_kind("udpc")
-}
-
-#[instrument(level = "debug")]
-pub fn udp_server_endpoints() -> Vec<String> {
-    get_endpoint_with_kind("udps")
-}
-
-#[instrument(level = "debug")]
-pub fn udp_broadcast_endpoints() -> Vec<String> {
-    get_endpoint_with_kind("udpb")
-}
-
-#[instrument(level = "debug")]
-pub fn serial_endpoints() -> Vec<String> {
-    get_endpoint_with_kind("serial")
-}
-
-#[instrument(level = "debug")]
-fn get_endpoint_with_kind(kind: &str) -> Vec<String> {
-    let mut endpoints = vec![];
-
-    for endpoint in MANAGER.clap_matches.endpoints.clone() {
-        let mut s = endpoint.split(':');
-
-        let Some(this_kind) = s.next() else {
-            warn!("No kind for endpoint: {endpoint:?}");
-            continue;
-        };
-
-        if !this_kind.eq(kind) {
-            trace!("ignoring: {this_kind:?}");
-            continue;
-        }
-
-        endpoints.push(s.clone().collect::<Vec<&str>>().join(":"));
-    }
-
-    endpoints
+pub fn endpoints() -> Vec<Arc<dyn drivers::Driver>> {
+    MANAGER.read().unwrap().clap_matches.endpoints.clone()
 }
 
-// Return the command line used to start this application
-
 #[instrument(level = "debug")]
 pub fn command_line_string() -> String {
     std::env::args().collect::<Vec<String>>().join(" ")
 }
 
-// Return a clone of current Args struct
-
-#[instrument(level = "debug")]
-pub fn command_line() -> String {
-    format!("{:#?}", MANAGER.clap_matches)
+#[cfg(test)]
+mod tests {
+    // Note this useful idiom: importing names from outer (for mod tests) scope.
+    use super::*;
+
+    #[test]
+    fn test_endpoints() {
+        let endpoints = vec![
+            "serial:/dev/ttyS0:9600",
+            "serial:COM1:115200",
+            "tcpc:10.0.0.1:4000",
+            "tcpc:localhost:7000",
+            "tcps:0.0.0.0:5000",
+            "tcps:192.168.1.10:6000",
+            "udpb:192.168.0.255:3000",
+            "udpb:255.255.255.255:9999",
+            "udpc:127.0.0.1:1234",
+            "udpc:192.168.1.100:8080",
+            "udps:0.0.0.0:5000",
+            "udps:192.168.1.5:6789",
+        ];
+
+        for endpoint in endpoints {
+            dbg!(endpoints_parser(endpoint).is_ok());
+        }
```

Comment:
I'm assuming this is a development tool, not a test finished test. A suggestion for test: having a Vec of `(endpoint: &str, driver: impl DriverExt)` and asserting each `endpoints_parser(endpoint)` call results in the expected Ok(driver).

Reply:
Check now